test/e2e/webhook_test.go

```diff
@@ -1049,3 +1052,305 @@ var _ = Describe("Fleet's Reserved Namespace Handler webhook tests", func() {
+				g.Expect(errors.As(err, &statusErr)).To(BeTrue(), fmt.Sprintf("Update cluster ressource binding call produced error %s. Error type wanted is %s.", reflect.TypeOf(err), reflect.TypeOf(&k8sErrors.StatusError{})))
+				g.Expect(string(statusErr.Status().Reason)).Should(Equal(fmt.Sprintf(resourceStatusErrFormat, testUser, testGroups, "ClusterResourceSnapshot", types.NamespacedName{Name: crs.Name})))
+				return nil
+			}, testutils.PollTimeout, testutils.PollInterval).Should(Succeed())
+		})
+
+		It("should deny DELETE cluster resource binding snapshot for user not in system:masters group", func() {
+			var crs placementv1beta1.ClusterResourceSnapshot
+			Expect(HubCluster.KubeClient.Get(ctx, types.NamespacedName{Name: "test-cluster-resource-snapshot"}, &crs)).Should(Succeed())
+			By("expecting denial of operation DELETE of cluster resource snapshot")
+			err := HubCluster.ImpersonateKubeClient.Delete(ctx, &crs)
+			var statusErr *k8sErrors.StatusError
+			Expect(errors.As(err, &statusErr)).To(BeTrue(), fmt.Sprintf("Delete cluster resource snapshot call produced error %s. Error type wanted is %s.", reflect.TypeOf(err), reflect.TypeOf(&k8sErrors.StatusError{})))
+			Expect(string(statusErr.Status().Reason)).Should(Equal(fmt.Sprintf(resourceStatusErrFormat, testUser, testGroups, "ClusterResourceSnapshot", types.NamespacedName{Name: crs.Name})))
+		})
+
+		It("should allow UPDATE operation on cluster resource snapshot for user in system:masters group", func() {
+			Eventually(func(g Gomega) error {
+				var crs placementv1beta1.ClusterResourceSnapshot
+				g.Expect(HubCluster.KubeClient.Get(ctx, types.NamespacedName{Name: "test-cluster-resource-snapshot"}, &crs)).Should(Succeed())
+				By("update labels in cluster resource snapshot")
+				labels := make(map[string]string)
+				labels[testKey] = testValue
+				crs.SetLabels(labels)
+
+				By("expecting successful UPDATE of cluster resource snapshot")
+				// The user associated with KubeClient is kubernetes-admin in groups: [system:masters, system:authenticated]
+				return HubCluster.KubeClient.Update(ctx, &crs)
+			}, testutils.PollTimeout, testutils.PollInterval).Should(Succeed())
+		})
+	})
+
+	Context("fleet guard rail ClusterSchedulingPolicySnapshot e2e tests", func() {
+		BeforeEach(func() {
+			placementPolicy := &placementv1beta1.PlacementPolicy{
+				PlacementType:    placementv1beta1.PickNPlacementType,
+				NumberOfClusters: pointer.Int32(3),
+				Affinity: &placementv1beta1.Affinity{
+					ClusterAffinity: &placementv1beta1.ClusterAffinity{
+						RequiredDuringSchedulingIgnoredDuringExecution: &placementv1beta1.ClusterSelector{
+							ClusterSelectorTerms: []placementv1beta1.ClusterSelectorTerm{
+								{
+									LabelSelector: metav1.LabelSelector{
+										MatchLabels: map[string]string{
+											"key1": "value1",
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+			}
+			jsonBytes, err := json.Marshal(placementPolicy)
```

Comment:
Hi Arvind! A nit here: for testing the hash calculation is really not necessary. The hash calculation part involves some additional logic (e.g., the num of clusters part is not included); it might be easier to just use an empty value.